pkg/controller/kubedirectorconfig/kubedirectorconfig_controller.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2018 BlueData Software, Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubedirectorconfig
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/bluek8s/kubedirector/pkg/apis/kubedirector.bluedata.io/v1alpha1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("controller_kubedirectorconfig")
+
+// Add creates a new KubeDirectorConfig Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileKubeDirectorConfig{Client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("kubedirectorconfig-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource KubeDirectorConfig
+	err = c.Watch(&source.Kind{Type: &v1alpha1.KubeDirectorConfig{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileKubeDirectorConfig implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileKubeDirectorConfig{}
+
+// ReconcileKubeDirectorConfig reconciles a KubeDirectorConfig object
+type ReconcileKubeDirectorConfig struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	Client client.Client
+	scheme *runtime.Scheme
+	lock         sync.RWMutex
```

Comment:
So originally we had one lock covering both clusterState and globalConfig. In this change, each of them has their own lock.

I think this is fine. Our only real consistency concerns were in making sure that an update to one of those items:
1) didn't allow a read of that same item until the update was done, and
2) didn't blow away an update to the other item that had just happened

As far as I know/remember there was no fundamental need for synchronizing access to both of them at the same time, from a semantic consistency standpoint. So I think this change is fine. Just talking it out here.

(I know you have other concerns about the right way to share access to the clusterState.)